[PECO-217] DBSQLClient.openSession refactoring

kravets-levko opened the pull request:
List of changes:

1. hide thrift details and expose only usable options (https://github.com/databricks/databricks-sql-nodejs/commit/0eb6634c81dee8614a7fbbd2ae1be0e78da5d7d5)
2. expose initial namespace config (https://github.com/databricks/databricks-sql-nodejs/commit/3b0b978980e0a4ebbd668c97778fdf42d4f2cc72)
3. use `client_protocol_i64` instead of `client_protocol` in order to enable Thrift protocol versions above `SPARK_CLI_SERVICE_PROTOCOL_V2` (https://github.com/databricks/databricks-sql-nodejs/commit/3b91837bed4394531d868c18245abdddb2d4d078)

yunbodeng-db commented on tests/unit/DBSQLClient.test.js:
```diff
@@ -62,24 +65,36 @@ describe('DBSQLClient.connect', () => {
-        expect(client.connectionProvider).instanceOf(HttpConnection);
-        cb();
-      })
-      .catch(cb);
```
About this change, it's just removing the meaningless test step? The test case calls cb() itself then verify cb() is invoked itself or I missed other significance of this change.

kravets-levko replied:
Code with `cb` is an old style of handling async operations in Mocha. You had to call `cb` without args to indicate that your test case succeeded and `cb` with error if you have unexpected error there. And the old code does so - `cb` is just called in `then()` and passed itself as handler for `catch` where it will accept the rejection error.

But with `async/await` all this becomes redundant - Mocha will handle everything properly. You just write your checks, and Mocha will catch all unexpected errors itself. I did this refactoring because it obviously makes code much simpler and more readable

yunbodeng-db commented on tests/unit/DBSQLClient.test.js:
```diff
@@ -91,23 +106,49 @@ describe('DBSQLClient.openSession', () => {
+    }
+  });
+
+  it('should throw an exception when the connection is lost', async () => {
```
This is almost same as the previous one - just different invalid connection status. If we have the third case, would you add another "complete" test case?
You could consolidate into a a base test method on given connection objects and the test cases themselves just pass in different invalid connections and expect the same behaviors.

it('should throw an exception when the connection is in an invalid state', async () => {
  for (<null_connection>, <disconnected_connection>) {
    verifyConnectionError(connection)
  }
}

verifyConnectionError () => {
}

kravets-levko replied:
Those test cases look similar, but they check a different branches in the code - when connection object wasn't created at all, and when it's available but not connected. For now both branches have similar behavior (throw an error), but I'd probably not merge these cases together. Especially considering that both are quite small, simple, and easy to read

yunbodeng-db commented on tests/unit/DBSQLClient.test.js:
```diff
@@ -119,47 +160,34 @@ describe('DBSQLClient.getClient', () => {
-      .catch(cb);
+
+    await client.close();
+    // No additional asserts needed - it should just reach this point
```
expect(closeConnectionStub.called).to.be.false?

yunbodeng-db replied:
Ditto below?

kravets-levko replied:
Tests not necessarily should contain checks :slightly_smiling_face: Sometimes absence of something is what we're expecting. This test case checks that `client.close()` should do nothing if connection wasn't initialized - so there's basically nothing to check except of that it doesn't fail - which is handled by Mocha

moderakh commented on lib/contracts/IDBSQLClient.ts:
```diff
@@ -1,18 +1,24 @@
 import IDBSQLSession from './IDBSQLSession';
 
-export interface IDBSQLConnectionOptions {
+export interface ConnectionOptions {
```
please add an entry in the changelog indicating that public API renamed
`IDBSQLConnectionOptions` -> `ConnectionOptions`

moderakh commented on lib/contracts/IDBSQLClient.ts:
```diff
@@ -1,18 +1,24 @@
+  initialCatalog?: string;
+  initialSchema?: string;
+  configuration?: Record<string, string>;
+  connectionProperties?: Record<string, string>;
```
as we adding this new API, we should understand what's the difference between `configuration` and `connectionProperties`. needs internal discussion

moderakh replied:
If we do not know the difference between `connectionProperties` and `configuration` we shouldn't make this a public API. (referring to our discussion in the internal google docs)

kravets-levko replied:
Yes, you're right. I'll figure this out, and either describe both options or remove them

moderakh commented on tests/unit/DBSQLClient.test.js:
```diff
@@ -62,24 +65,36 @@ describe('DBSQLClient.connect', () => {
+    const client = new DBSQLClient();
+    client.client = {
+      OpenSession(req, cb) {
+        cb(null, { status: {}, sessionHandle: {} });
```
why do we need a mix of async and callback `cb` testing?

can't we write the tests to use async instead of callbacks?

kravets-levko replied:
This callback is not Mocha's - it's just a part of the mock object (ThriftClient uses callback-based interface)

moderakh commented on lib/contracts/IDBSQLClient.ts:
```diff
@@ -1,18 +1,24 @@
 
+export interface OpenSessionRequest {
+  initialCatalog?: string;
+  initialSchema?: string;
```
these are nullable right? if we don't set the initial catalog and initial schema, what happens? do we bind to hive-metastore?

kravets-levko replied:
Yes, server will set something by default. So far it always was `hive_metastore.default`

moderakh reviewed (dismissed):
Thanks @kravets-levko except the following looks good to me.

The PR is adding `OpenSessionRequest` and exposes `configuration` and `connectionProperties` we should discuss internally if these to be made part of the public API. if we are not sure, it is better to not make it public till we decide (otherwise changing that later will be a breaking change)

Please also high level keep track of the breaking changes in the public API and list them in the changelog. (ideally if we could update the changelog in the same PR would be better to make sure we don't forget)

kravets-levko commented:
@moderakh it turns out that things are not clear about `configuration` and `connectionProperties` options, so I removed them for now. Also, I updated changelog - added changes from this PR and also from previous one (forgot to do it there)

moderakh reviewed (approved):
LGTM. Thanks @kravets-levko great work and thanks for addressing the comments.
